File: src/Features/Redis_LockKey.php
<?php

namespace xltxlm\redis\Features;


use xltxlm\redis\Exception\Redis_LockKey\Expire_Exception;
use xltxlm\redis\Exception\Redis_LockKey\LockFall_Exception;

/**
 * 锁定函数;
 */
class Redis_LockKey extends Redis_LockKey\Redis_LockKey_implements
{
    /**
     * Redis_LockKey constructor.
     */
    public function __construct(string $key = '', int $expire = 0)
    {
        if ($key) {
            $this->setkey($key);
        }
        if ($expire) {
            $this->setexpire($expire);
        }
    }

    public function __invoke(): bool
    {
        if ($this->getExpire() < 1) {
            throw new Expire_Exception();
        }
        /** @var \Redis $redis */
        $redis = $this->getRedisConfig()->__invoke();
        $waittimes = 0;
        do {
            //写入key,并且设置过期时间
            //设置上锁的时候，同时也给key设置一个过期时间，防止没有解锁，导致以后再也取不到锁了。
            if ($redis->set($this->getKey(), $this->getValue(), ['nx', 'ex' => $this->getExpire()])) {
                return true;
            }

            //锁定失败的时候抛出异常
            if ($this->getException_on_LockFail()) {
                throw (new LockFall_Exception())
                    ->setRedisConfig($this->getRedisConfig())
                    ->setRedis_LockKey($this);
            }
            //获取锁失败的时候，可以暂停
            if ($this->getTry_Wait_Second()) {
                $waittimes++;
                if ($waittimes > $this->getTry_Wait_Second()) {
                    return false;
                }
                //暂停1秒
                sleep(1);
            } else {
                return false;
            }
        } while (true);
    }

    /**
     * 使用内置的lua脚本来解锁。因为需要事务逻辑
     *
     * @return bool
     */
    public function Free(): bool
    {
        /** @var \Redis $redis */
        $redis = $this->getRedisConfig()->__invoke();
        return $redis->eval("if redis.call('get', KEYS[1]) == ARGV[1] then return redis.call('del', KEYS[1]) else return 0 end", [$this->getKey(), $this->getValue()], 1);
    }


}

File: go/Features/Redis_LockKey.go
package Features

type Redis_LockKey struct {
    /* 锁定的key */
    key string

    /* 超时时间 */
    expire int

    /* 服务器配置 */
    RedisConfig 

    /* 需要存储的值,默认是当前时间 */
    value string

    /* 如果一次锁不住,希望卡住坚持的秒数 */
    Try_Wait_Second int

    /* 锁定失败的时候,抛出异常 */
    Exception_on_LockFail bool

}

func NewRedis_LockKey(RedisConfig ,Try_Wait_Second int,Exception_on_LockFail bool) *Redis_LockKey{
    var this = new(Redis_LockKey)
    this.SetRedisConfig(RedisConfig);
    this.SetTry_Wait_Second(Try_Wait_Second);
    this.SetException_on_LockFail(Exception_on_LockFail);
    return this
}

func (this *Redis_LockKey)GetRedisConfig() {
    return this.RedisConfig;
}

func (this *Redis_LockKey)SetRedisConfig(RedisConfig ) *Redis_LockKey{
    this.RedisConfig = RedisConfig;
    return this
}
func (this *Redis_LockKey)GetTry_Wait_Second() int{
    return this.Try_Wait_Second;
}

func (this *Redis_LockKey)SetTry_Wait_Second(Try_Wait_Second int) *Redis_LockKey{
    this.Try_Wait_Second = Try_Wait_Second;
    return this
}
func (this *Redis_LockKey)GetException_on_LockFail() bool{
    return this.Exception_on_LockFail;
}

func (this *Redis_LockKey)SetException_on_LockFail(Exception_on_LockFail bool) *Redis_LockKey{
    this.Exception_on_LockFail = Exception_on_LockFail;
    return this
}

/**
    检测是否已经锁住了*/
func (this *Redis_LockKey)__invoke()bool{

}

/**
    执行的是事务的时候,还需要主动释放掉key*/
func (this *Redis_LockKey)Free()bool{

}

